### PR TITLE
add gpu device type

### PR DESCRIPTION
**Pull request by brauner**
- lxc config device add c1 gpu0 gpu => Pass everything we have
- lxc config device add c1 gpu0 gpu vendor=10de => Pass whatever we have from nvidia
- lxc config device add c1 gpu0 gpu vendor=10de id=1 => Pass the second nvidia GPU
- lxc config device add c1 gpu0 gpu pci=0000:02:08.0 => Pass whatever GPU is at that PCI address

This parses /sys/bus/pci/devices/* and looks for the drm subfolder. All
subfolders under that, e.g.

	card0
	controlD64
	renderD128

belong to the same PCI address and probably all need to be handed to the
container when the user somehow selects card0.
We then retrieve the vendor ID, and if available the device ID from

	/sys/bus/pci/devices/*/vendor
	/sys/bus/pci/devices/*/device

The major and minor number for each device can be gathered from the

	card0/dev
	controlD64/dev
	renderD128/dev

files.

If we detect that the vendor ID is Nvidia, we also parse /dev and look for

	/dev/nvidia[0-9]+

entries. We thereby associate the correct /dev/nvidia* entry with the correct
/dev/card* entry, e.g. we correlate

	/dev/card0
and

    /dev/nvidia0.

Finally, we store any residual /dev/nvidia[^0-9]+ mounts that are not cards
because we need to assume that if /dev/nvidia[0-9]+ is requested then the other
files are necessary for correct functionality of that card and need to be
mounted into every container that also mounts that card.

Signed-off-by: Christian Brauner <christian.brauner@canonical.com>

### CODE REVIEW

**Comment by brauner** [comment]
Closes #2535.

**Comment by brauner** [comment]
Bind-mounted `Nvidia` test tools:
```
ubuntu@canonical-lxd:~$ ./lxc config show gpu
name: gpu
profiles:
- default
config:
  volatile.base_image: 6f25adef061c3f2186c6910bff8cacd0c63e1493e3f8b616e52eb84076890bd1
  volatile.eth0.hwaddr: 00:16:3e:b9:18:6c
  volatile.last_state.idmap: '[{"Isuid":true,"Isgid":false,"Hostid":165536,"Nsid":0,"Maprange":65536},{"Isuid":false,"Isgid":true,"Hostid":165536,"Nsid":0,"Maprange":65536}]'
  volatile.last_state.power: RUNNING
devices:
  nvidia:
    path: /home/ubuntu/NVIDIA_CUDA-7.5_Samples/
    source: /home/ubuntu/NVIDIA_CUDA-7.5_Samples/
    type: disk
  root:
    path: /
    type: disk
ephemeral: false
```

Add `NVIDIA GPU 1`:
```
ubuntu@canonical-lxd:~$ ./lxc config device add gpu gpu0 gpu vendorid=10de id=1
Device gpu0 added to gpu
```

Confirm it's available in the container:
```
ubuntu@canonical-lxd:~$ ./lxc exec gpu -- ls -al /dev/dri/; ls -al /dev/ | grep nvidia
total 0
drwxr-xr-x  2 root root     60 Oct 20 14:04 .
drwxr-xr-x 10 root root    600 Oct 20 14:04 ..
crw-rw----  1 root root 226, 1 Oct 20 14:04 card1
crw-rw-rw-  1 root   root    195,   0 Oct 17 18:59 nvidia0
crw-rw-rw-  1 root   root    195,   1 Oct 17 18:59 nvidia1
crw-rw-rw-  1 root   root    195, 255 Oct 17 18:59 nvidiactl
crw-rw-rw-  1 root   root    246,   0 Oct 17 19:00 nvidia-uvm
crw-rw-rw-  1 root   root    246,   1 Oct 17 19:10 nvidia-uvm-tools
```

Run test:
```
ubuntu@canonical-lxd:~$ ./lxc exec gpu -- /home/ubuntu/NVIDIA_CUDA-7.5_Samples/bin/x86_64/linux/release/matrixMulCUBLAS
[Matrix Multiply CUBLAS] - Starting...
GPU Device 0: "GeForce GT 730" with compute capability 3.5

MatrixA(640,480), MatrixB(480,320), MatrixC(640,320)
Computing result using CUBLAS...done.
Performance= 274.25 GFlop/s, Time= 0.717 msec, Size= 196608000 Ops
Computing result using host CPU...done.
Comparing CUBLAS Matrix Multiply with CPU results: PASS

NOTE: The CUDA Samples are not meant for performance measurements. Results may vary when GPU Boost is enabled.
```

Add `NVIDIA GPU 0` while running
```
ubuntu@canonical-lxd:~$ ./lxc config device add gpu gpu1 gpu vendorid=10de id=0
Device gpu1 added to gpu
```

Confirm it's there
```
ubuntu@canonical-lxd:~$ ./lxc exec gpu -- ls -al /dev/dri/; ls -al /dev/ | grep nvidia
total 0
drwxr-xr-x  2 root root     80 Oct 20 14:09 .
drwxr-xr-x 10 root root    620 Oct 20 14:09 ..
crw-rw----  1 root root 226, 0 Oct 20 14:09 card0
crw-rw----  1 root root 226, 1 Oct 20 14:04 card1
crw-rw-rw-  1 root   root    195,   0 Oct 17 18:59 nvidia0
crw-rw-rw-  1 root   root    195,   1 Oct 17 18:59 nvidia1
crw-rw-rw-  1 root   root    195, 255 Oct 17 18:59 nvidiactl
crw-rw-rw-  1 root   root    246,   0 Oct 17 19:00 nvidia-uvm
crw-rw-rw-  1 root   root    246,   1 Oct 17 19:10 nvidia-uvm-tools
```

File: doc/configuration.md [hunk 1]
```diff
@@ -277,6 +277,20 @@ gid         | int       | 0                 | no        | GID of the device owne
...
 
+### Type: gpu
+GPU device entries simply make the requested gpu device appear in the
+container.
+
+The following properties exist:
+
+Key         | Type      | Default           | Required  | Description
+:--         | :--       | :--               | :--       | :--
+vendorid    | string    | -                 | no        | The vendor id of the GPU device.
```
Comment (stgraber) [inline doc/configuration.md]: I guess we may as well add productid just to match USB, obviously neither would be required.

**Comment by stgraber** [comment]
Missing a doc/api-extension.md entry

**Comment by brauner** [comment]
> Missing a doc/api-extension.md entry

I was wondering about that since we do not have one for `usb_devices`. I'll add that too.

File: doc/configuration.md (hunk 1 above)
Comment (brauner) [inline doc/configuration.md]: Addressed.

**Comment by stgraber** [comment]
Oh oops, must have missed that when I reviewed the USB stuff. Bad @tych0 ! ;)

File: doc/configuration.md (hunk 1 above)
Comment (stgraber) [inline doc/configuration.md]: Sorry, I meant "in addition" to the "id" field as they don't have the same meaning.

id is a card id, so 0 for the first card, 1 for the second, ... that matches the DRI or NVidia card id.

productid refers to the PCI productid

So productid=1234 may match more than one card if you have more than one card of the same model, vendor=abcd  & id=1 will only ever match a single card.

File: doc/api-extensions.md [hunk 2]
```diff
@@ -65,6 +65,9 @@ Add support for the HTTP PATCH method.
 
 PATCH allows for partial update of an object in place of PUT.
 
+## usb\_devices
+Add support for USB hotplug.
+
```
Comment (stgraber) [inline doc/api-extensions.md]: Can we get that in a separate commit so I can put it in the 2.4.2 bugfix release?

Comment (brauner) [inline doc/api-extensions.md]: Oh, sure. :)

**Comment by brauner** [comment]
I think `/lxd/build/tmp.PnUmITxUTh/go/bin/lxc restart foo --force --debug` failed due to a `lxc stop` timeout and the other one is `/lxd/build/tmp.m3ivmWREqZ/go/bin/lxc move l1:nonlive l2: --debug`. Both seem unrelated to this patch.

**Comment by stgraber** [comment]
Yep, so that leaves the vendorid/id thing to address then.

**Comment by brauner** [comment]
That I did, I think. Changed it to vendorid and productid.

File: doc/configuration.md (hunk 1 above)
Comment (brauner) [inline doc/configuration.md]: Oh this. Yes, will do. :)

**Comment by brauner** [comment]
There's a refcount issue with respect to the additional `nvidia` mounts I need to think about.

**Comment by brauner** [comment]
So I had to find a way to guarantee that additional mounts for nvidia cards are present until the last card has been removed. The last comment should do the trick. Also addressed the `productid` and `id` issue.

**Comment by brauner** [comment]
I think jenkins went offline on some builders.

**Comment by stgraber** [comment]
Jenkins should be fine, North American internet wasn't doing so great today with Dyn going down and taking Github and a bunch of other sites (including ubuntu.com) with it for a while.

**Comment by stgraber** [comment]
jenkins: test this please

**Pull request merged**